[Python] querying articles batch 1 [Replace C# samples] (#1797)
ravendb/docs: Documentation/6.0/Raven.Documentation.Pages/client-api/session/querying/sort-query-results.python.markdown

 
    * [Order by score](../../../client-api/session/querying/sort-query-results#order-by-score)
        * [Get resulting score](../../../client-api/session/querying/sort-query-results#get-resulting-score)
  
    * [Order by random](../../../client-api/session/querying/sort-query-results#order-by-random)
   
    * [Order by spatial](../../../client-api/session/querying/sort-query-results#order-by-spatial)
     
    * [Order by count (aggregation query)](../../../client-api/session/querying/sort-query-results#order-by-count-(aggregation-query))
  
    * [Order by sum (aggregation query)](../../../client-api/session/querying/sort-query-results#order-by-sum-(aggregation-query))

    * [Force ordering type](../../../client-api/session/querying/sort-query-results#force-ordering-type)

    * [Chain ordering](../../../client-api/session/querying/sort-query-results#chain-ordering)

    * [Custom sorters](../../../client-api/session/querying/sort-query-results#custom-sorters) 

    * [Syntax](../../../client-api/session/querying/sort-query-results#syntax)

{NOTE/}

---

{PANEL: Order by field value}

> * Use `OrderBy` or `OrderByDescending` to order the results by the specified document-field.

Reviewer: ```suggestion
* Use `order_by` or `order_by_descending` to order the results by the specified document-field.
```

Author: seems like you're reviewing an old commit? these are already fixed

{CODE-TABS}
{CODE-TAB:python:Query sort_1@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> where UnitsInStock > 10

Reviewer: ```suggestion
where units_in_stock > 10
```

Author: already fixed

> order by UnitsInStock as long

Reviewer: ```suggestion
order by units_in_stock as long
```

Author: already fixed

{CODE-TAB-BLOCK/}
{CODE-TABS/}

{INFO: }

**Ordering Type**:

* By default, the `order_by` methods will determine the `OrderingType` from the property path expression  
  and specify that ordering type in the generated RQL that is sent to the server.  

* E.g. in the above example, ordering by `x => x.units_in_stock` will result in `OrderingType.LONG`  
  because that property data type is an integer.

* Different ordering can be forced - see [Force ordering type](../../../client-api/session/querying/sort-query-results#force-ordering-type) below.

{INFO/}

{PANEL/}

{PANEL: Order by score}

* When querying with some filtering conditions, a basic score is calculated for each item in the results  
  by the underlying indexing engine.

* The higher the score value the better the match.  

* Use `order_by_score` or `order_by_score_descending` to order the query results by this score.

{CODE-TABS}
{CODE-TAB:python:Query sort_4@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> where UnitsInStock < 5 or Discontinued == true

Author: already fixed

  Learn more in [Include query explanations](../../../client-api/session/querying/debugging/include-explanations).

* **Get score from metadata**:    

    * The score is available in the `@index-score` metadata property within each result.  
      Note the following difference between the underlying indexing engines:

      * When using **Lucene**:  
          This metadata property is always available in the results.  
          Read more about Lucene scoring [here](https://lucene.apache.org/core/3_3_0/scoring.html).

      * When using **Corax**:  
          In order to enhance performance, this metadata property is Not included in the results by default.  
          To get this metadata property you must set the [Indexing.Corax.IncludeDocumentScore](../../../server/configuration/indexing-configuration#indexing.corax.includedocumentscore) configuration value to `True`  
          Learn about the available methods for setting an indexing configuration key in this [indexing-configuration](../../../server/configuration/indexing-configuration) article.

{INFO/}

{PANEL/}

{PANEL: Order by random}

> * Use `RandomOrdering` to randomize the order of the query results.

Reviewer: ```suggestion
* Use `random_ordering` to randomize the order of the query results.
```

Author: already fixed

* An optional seed parameter can be passed.

{CODE-TABS}
{CODE-TAB:python:Query sort_7@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> where UnitsInStock > 10

Reviewer: ```suggestion
where units_in_stock > 10
```

Author: already fixed

order by random()
// order by random(someSeed)
{CODE-TAB-BLOCK/}
{CODE-TABS/}

{PANEL/}

{PANEL: Order by spatial}

* If your data contains geographical locations,  
  spatial query results can be sorted based on their distance from a specific point.

* See detailed explanation in [Spatial Sorting](../../../client-api/session/querying/how-to-make-a-spatial-query#spatial-sorting).

{PANEL/}

{PANEL: Order by count (aggregation query)}

* The results of a [group_by query](../../../client-api/session/querying/how-to-perform-group-by-query) can be sorted by the `count` aggregation operation used in the query.

{CODE-TABS}
{CODE-TAB:python:Query sort_10@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> group by Category

Reviewer: ```suggestion
group by category
```

order by count() as long

> select key() as "Category", count()

Reviewer: ```suggestion
select key() as "category", count()
```

Author: done

{CODE-TAB-BLOCK/}
{CODE-TABS/}

{PANEL/}

{PANEL: Order by sum (aggregation query)}

* The results of a [group_by query](../../../client-api/session/querying/how-to-perform-group-by-query) can be sorted by the `sum` aggregation operation used in the query.

{CODE-TABS}
{CODE-TAB:python:Query sort_13@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> group by Category

Reviewer: ```suggestion
group by category
```

order by Sum as long

> select key() as 'Category', sum(UnitsInStock) as Sum

Reviewer: ```suggestion
select key() as 'category', sum(units_in_stock) as Sum
```

Author: done

{PANEL: Force ordering type}

* By default, the `order_by` methods will determine the `OrderingType` from the property path expression  
  and specify that ordering type in the generated RQL that is sent to the server.

* A different ordering can be forced by passing the ordering type explicitly to `order_by` or `order_by_descending`.

* The following ordering types are available:

    * `OrderingType.LONG`
    * `OrderingType.DOUBLE`
    * `OrderingType.ALPHA_NUMERIC`
    * `OrderingType.STRING` (lexicographic ordering)

* When using RQL directly, if no ordering type is specified, then the server defaults to lexicographic ordering.

{NOTE: }

**Using alphanumeric ordering example**:

* When ordering mixed-character strings by the default lexicographical ordering  
  then comparison is done character by character based on the Unicode values.  
  For example, "Abc9" will come after "Abc10" since 9 is greater than 1.

* If you want the digit characters to be ordered as numbers then use alphanumeric ordering  
  where "Abc10" will result after "Abc9".

{CODE-TABS}
{CODE-TAB:python:Query sort_16@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> order by QuantityPerUnit as alphanumeric

Reviewer: ```suggestion
order by quantity_per_unit as alphanumeric
```

Author: already fixed

{CODE-TAB-BLOCK/}
{CODE-TABS/}

{NOTE/}

{PANEL/}

{PANEL: Chain ordering}

* It is possible to chain multiple orderings in the query.  
  Any combination of secondary sorting is possible as the fields are indexed independently of one another.

* There is no limit on the number of sorting actions that can be chained.
  
* This is achieved by using the `then_by` (`then_by_descending`) and `then_by_score` (`then_by_score_descending`) methods.

{CODE-TABS}
{CODE-TAB:python:Query sort_19@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> where UnitsInStock > 10

Reviewer: ```suggestion
where units_in_stock > 10
```

Author: already fixed

> order by UnitsInStock as long desc, score(), Name

Reviewer: ```suggestion
order by units_in_stock as long desc, score(), Name
```

Author: already fixed

{CODE-TAB-BLOCK/}
{CODE-TABS/}

{PANEL/}

{PANEL: Custom sorters }

* The Lucene indexing engine allows you to create your own custom sorters.  
  Custom sorters are not supported by [Corax](../../../indexes/search-engine/corax).  
 
* Custom sorters can be deployed to the server by either:  

     * Sending the [Put Sorters Operation](../../../client-api/operations/maintenance/sorters/put-sorter) from your code.
  
     * Uploading a custom sorter from Studio, see [Custom Sorters View](../../../studio/database/settings/custom-sorters).

* Once the custom sorter is deployed, you can sort the query results with it.

{CODE-TABS}
{CODE-TAB:python:Query sort_22@ClientApi\Session\Querying\SortQueryResults.py /}
{CODE-TAB-BLOCK:sql:RQL}
from "Products"

> where UnitsInStock > 10

Reviewer: ```suggestion
where units_in_stock > 10
```

> order by custom(UnitsInStock, "MySorter")

Reviewer: ```suggestion
order by custom(units_in_stock, "MySorter")
```

Author: already fixed
